File: src/main/java/com/github/kuro46/embedscript/util/Scheduler.java
package com.github.kuro46.embedscript.util;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * @author shirokuro
 */
public final class Scheduler {
    private static final ExecutorService EXECUTOR = Executors.newCachedThreadPool(r -> {
        Thread thread = new Thread(r);
        thread.setName("EmbedScript Scheduler Thread");
        thread.setDaemon(true);
        thread.setUncaughtExceptionHandler((t, e) -> {
            System.err.println("Exception in " + t.getName());
            e.printStackTrace();
        });

        return thread;
    });

    private Scheduler() {
    }

    public static void execute(Runnable task) {
        EXECUTOR.execute(task);
    }

    public static Future<?> submit(Runnable task) {
        return EXECUTOR.submit(task);
    }

    public static <T> Future<T> submit(Callable<T> task) {
        return EXECUTOR.submit(task);
    }

    public static <T> Future<T> submit(Runnable task, T t) {
        return EXECUTOR.submit(task, t);
    }
}

File: src/main/kotlin/com/github/kuro46/embedscript/util/Scheduler.kt
package com.github.kuro46.embedscript.util

import com.google.common.util.concurrent.Futures
import java.util.Collections
import java.util.WeakHashMap
import java.util.concurrent.Executors
import java.util.concurrent.Future

/**
 * @author shirokuro
 */
object Scheduler {
    private val threads: MutableSet<Thread> = Collections.synchronizedSet(Collections.newSetFromMap(WeakHashMap()))
    private val executor = Executors.newCachedThreadPool { r ->
        val thread = Thread(r)
        thread.name = "EmbedScript Scheduler Thread"
        thread.isDaemon = true

        threads.add(thread)

        thread
    }

    fun execute(forceOtherThread: Boolean = false, task: () -> Unit) {
        if (!forceOtherThread && threads.contains(Thread.currentThread())) {
            task()
        } else {
            executor.execute(wrapFunction(task))
        }
    }

    fun submit(forceOtherThread: Boolean = false, task: () -> Unit): Future<*> {
        return if (!forceOtherThread && threads.contains(Thread.currentThread())) {
            Futures.immediateFuture(null)
        } else {
            executor.submit(wrapFunction(task))
        }
    }

    private fun wrapFunction(task: () -> Unit): () -> Unit {
        return {
            try {
                task()
            } catch (e: Throwable) {
                System.err.println("Exception or Error occurred in '${Thread.currentThread().name}' !")
                e.printStackTrace()
            }
        }
    }
}
